source/lib.ts

```diff
@@ -1,51 +1,63 @@
 import {queryAdditionalPermissions} from 'webext-permissions';
 import {onExtensionStart} from 'webext-events';
-import {excludeDuplicateFiles} from './deduplicator.js';
 import {injectToExistingTabs} from './inject-to-existing-tabs.js';
-import {registerContentScript} from './register-content-script-shim.js';
 
-const registeredScripts = new Map<
-string,
-Promise<browser.contentScripts.RegisteredContentScript>
->();
+type ManifestContentScript = ReturnType<typeof getContentScripts>[number];
 
 // In Firefox, paths in the manifest are converted to full URLs under `moz-extension://` but browser.contentScripts expects exclusively relative paths
 function makePathRelative(file: string): string {
 	return new URL(file, location.origin).pathname;
 }
 
 function getContentScripts() {
-	const {content_scripts: rawManifest, manifest_version: manifestVersion} = chrome.runtime.getManifest();
+	const {content_scripts: scripts} = chrome.runtime.getManifest();
 
-	if (!rawManifest) {
-		throw new Error('webext-dynamic-content-scripts tried to register scripts on the new host permissions, but no content scripts were found in the manifest.');
+	if (scripts) {
+		return scripts;
 	}
 
-	return excludeDuplicateFiles(rawManifest, {warn: manifestVersion === 2});
+	throw new Error('webext-dynamic-content-scripts tried to register scripts on the new host permissions, but no content scripts were found in the manifest.');
+}
+
+function getIdFromConfig(origin: string, config: ManifestContentScript) {
+	return 'webext-dynamic-content-script-' + JSON.stringify({...config, matches: [origin]});
+}
+
+async function register(origin: string, config: ManifestContentScript) {
+	try {
+		await chrome.scripting.registerContentScripts([{
+			id: getIdFromConfig(origin, config),
+			// Always convert paths here because we don't know whether Firefox MV3 will accept full URLs
+			js: config.js?.map(file => makePathRelative(file)),
+			css: config.css?.map(file => makePathRelative(file)),
+			allFrames: config.all_frames,
+			matches: [origin],
+			excludeMatches: config.matches,
+			persistAcrossSessions: true,
+			runAt: config.run_at as browser.extensionTypes.RunAt,
+		}]);
+	} catch (error) {
+		if ((error as Error)?.message.startsWith('Duplicate script ID')) {
+			// Previously registered, nothing to do
+			return;
+		}
+
+		// Unknown error, throw
+		throw error;
+	}
+
+	// Registration successful, now inject into existing tabs
+	await injectToExistingTabs([origin], [config]);
```

Comment:
Ideally the browser still remembers scripts registered in past sessions, so this this now only runs when the scripts _were not_ previously registered.